README.md
# arabic-roman-convert.js

[![Commitizen friendly][commitizen-image]][commitizen-url]
[![XO code style][codestyle-image]][codestyle-url]

[![NPM version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Coveralls Status][coveralls-image]][coveralls-url]
[![Dependency Status][depstat-image]][depstat-url]
[![DevDependency Status][depstat-dev-image]][depstat-dev-url]

> Convert [Arabic][wiki-arabic] to [Roman][wiki-roman] UTF-8 numerals: 7 → Ⅶ

Demo: [vovanr.github.io/arabic-roman-convert.js][demo]

**Arabic numerals** `.arabicToRoman({Number|String})`
```
1 → Ⅰ
2 → Ⅱ
3 → Ⅲ
4 → Ⅳ
5 → Ⅴ
6 → Ⅵ
7 → Ⅶ
8 → Ⅷ
9 → Ⅸ
10 → Ⅹ
11 → Ⅺ
12 → Ⅻ
15 → ⅩⅤ
25 → ⅩⅩⅤ
50 → Ⅼ
100 → Ⅽ
500 → Ⅾ
1000 → Ⅿ
```

**Simple roman strings** `.convertRoman({String})`
```
I → Ⅰ
III → Ⅲ
vii → Ⅶ
ix → Ⅸ
X → Ⅹ
XV → ⅩⅤ
l → Ⅼ
c → Ⅽ
D → Ⅾ
M → Ⅿ
```

**Arabic number or simple Roman string** `.toRoman({String|Number})`
```
2 → Ⅱ
III → Ⅲ
vii → Ⅶ
9 → Ⅸ
X → Ⅹ
XV → ⅩⅤ
l → Ⅼ
100 → Ⅽ
D → Ⅾ
M → Ⅿ
```


## Install and Usage

### Bower
```sh
bower install --save arabic-roman-convert.js
```
```html
<script src="bower_components/arabic-roman-convert.js/dist/arabic-roman-convert.min.js"></script>
```
```js
var num = window.ArabicRoman.toRoman(2016);
//=> ⅯⅯⅩⅥ
```

### Node
```sh
npm install --save arabic-roman-convert.js
```
```js
var arabicRomanConvert = require('arabic-roman-convert.js');

var num = arabicRomanConvert.toRoman(2016);
//=> ⅯⅯⅩⅥ
```

## License
MIT © [Vladimir Rodkin](https://github.com/VovanR)

[wiki-arabic]: http://en.wikipedia.org/wiki/Arabic_numerals
[wiki-roman]: http://en.wikipedia.org/wiki/Roman_numerals
[demo]: https://vovanr.github.io/arabic-roman-convert.js

[commitizen-url]: http://commitizen.github.io/cz-cli/
[commitizen-image]: https://img.shields.io/badge/commitizen-friendly-brightgreen.svg?style=flat-square

[codestyle-url]: https://github.com/sindresorhus/xo
[codestyle-image]: https://img.shields.io/badge/code_style-XO-5ed9c7.svg?style=flat-square

[npm-url]: https://npmjs.org/package/arabic-roman-convert.js
[npm-image]: http://img.shields.io/npm/v/arabic-roman-convert.js.svg?style=flat-square

[travis-url]: https://travis-ci.org/VovanR/arabic-roman-convert.js
[travis-image]: http://img.shields.io/travis/VovanR/arabic-roman-convert.js.svg?style=flat-square

[coveralls-url]: https://coveralls.io/r/VovanR/arabic-roman-convert.js
[coveralls-image]: http://img.shields.io/coveralls/VovanR/arabic-roman-convert.js.svg?style=flat-square

[depstat-url]: https://david-dm.org/VovanR/arabic-roman-convert.js
[depstat-image]: https://david-dm.org/VovanR/arabic-roman-convert.js.svg?style=flat-square

[depstat-dev-url]: https://david-dm.org/VovanR/arabic-roman-convert.js
[depstat-dev-image]: https://david-dm.org/VovanR/arabic-roman-convert.js/dev-status.svg?style=flat-square


CHANGELOG.md
<a name="1.2.1"></a>
## [1.2.1](https://github.com/vovanr/arabic-roman-convert.js/compare/v1.2.0...v1.2.1) (2016-04-25)






test.js
import test from 'ava';
import sinon from 'sinon';
import arabicRoman from './';

const cod = {
	I: '\u2160',
	II: '\u2161',
	III: '\u2162',
	IV: '\u2163',
	V: '\u2164',
	VI: '\u2165',
	VII: '\u2166',
	VIII: '\u2167',
	IX: '\u2168',
	X: '\u2169',
	XI: '\u216A',
	XII: '\u216B',
	L: '\u216C',
	C: '\u216D',
	D: '\u216E',
	M: '\u216F'
};

/**
 * @return {String}
 */
const charToCode = function () {
	let res = '';
	const args = arguments;
	let i;
	for (i = 0; i < args.length; i++) {
		res += cod[args[i]];
	}
	return res;
};

// toRoman
test('#toRoman should fire `arabicToRoman` if argument is arabic number', t => {
	const spy = sinon.spy(arabicRoman, 'arabicToRoman');

	arabicRoman.toRoman(3);
	arabicRoman.toRoman('5');

	t.true(spy.called);
	t.is(spy.getCall(0).args[0], 3);
	t.is(spy.getCall(1).args[0], '5');

	spy.restore();
});

test('#toRoman should fire `convertRoman` if argument is simple roman', t => {
	const spy = sinon.spy(arabicRoman, 'convertRoman');

	arabicRoman.toRoman('III');

	t.true(spy.called);
	t.is(spy.getCall(0).args[0], 'III');

	spy.restore();
});

// arabicToRoman
test('#arabicToRoman should not convert number < 1', t => {
	t.is(arabicRoman.arabicToRoman(0), undefined);
	t.is(arabicRoman.arabicToRoman(-1), undefined);
});

test('#arabicToRoman should not convert nubmer > 4000', t => {
	t.truthy(arabicRoman.arabicToRoman(4000));
	t.is(arabicRoman.arabicToRoman(4001), undefined);
});

test('#arabicToRoman should parse string attributes', t => {
	t.is(arabicRoman.arabicToRoman('1'), '\u2160');
	t.is(arabicRoman.arabicToRoman(NaN), undefined);
	t.is(arabicRoman.arabicToRoman('a1'), undefined);
});

test('#arabicToRoman should convert arabic to roman numerals (from 1 to 10)', t => {
	const codes = [
		'',
		charToCode('I'),
		charToCode('II'),
		charToCode('III'),
		charToCode('IV'),
		charToCode('V'),
		charToCode('VI'),
		charToCode('VII'),
		charToCode('VIII'),
		charToCode('IX'),
		charToCode('X')
	];
	for (let i = 1; i < codes.length - 1; i++) {
		t.is(arabicRoman.arabicToRoman(i), codes[i]);
	}
});

test('#arabicToRoman should convert big numbers', t => {
	t.is(arabicRoman.arabicToRoman(11), charToCode('XI'));
	t.is(arabicRoman.arabicToRoman(12), charToCode('XII'));
	t.is(arabicRoman.arabicToRoman(50), charToCode('L'));
	t.is(arabicRoman.arabicToRoman(100), charToCode('C'));
	t.is(arabicRoman.arabicToRoman(500), charToCode('D'));
	t.is(arabicRoman.arabicToRoman(1000), charToCode('M'));
});

test('#arabicToRoman should convert complex numbers', t => {
	t.is(arabicRoman.arabicToRoman(13), charToCode('X', 'III'));
	t.is(arabicRoman.arabicToRoman(14), charToCode('X', 'IV'));
	t.is(arabicRoman.arabicToRoman(15), charToCode('X', 'V'));
	t.is(arabicRoman.arabicToRoman(19), charToCode('X', 'IX'));
	t.is(arabicRoman.arabicToRoman(20), charToCode('X', 'X'));
	t.is(arabicRoman.arabicToRoman(39), charToCode('X', 'X', 'X', 'IX'));
	t.is(arabicRoman.arabicToRoman(4000), charToCode('M', 'M', 'M', 'M'));
	t.is(
		arabicRoman.arabicToRoman(3888),
		charToCode('M', 'M', 'M', 'D', 'C', 'C', 'C', 'L', 'X', 'X', 'X', 'VIII')
	);
	t.is(arabicRoman.arabicToRoman(3999), charToCode('M', 'M', 'M', 'C', 'M', 'X', 'C', 'IX'));
});

test('#arabicToRoman should convert hard numbers', t => {
	t.is(arabicRoman.arabicToRoman(4), charToCode('IV'));
	t.is(arabicRoman.arabicToRoman(9), charToCode('IX'));
	t.is(arabicRoman.arabicToRoman(40), charToCode('X', 'L'));
	t.is(arabicRoman.arabicToRoman(90), charToCode('X', 'C'));
	t.is(arabicRoman.arabicToRoman(400), charToCode('C', 'D'));
	t.is(arabicRoman.arabicToRoman(900), charToCode('C', 'M'));
});

// convertRoman
test('#convertRoman should parse string attributes', t => {
	t.is(arabicRoman.convertRoman(3), undefined);
	t.is(arabicRoman.convertRoman('asdif'), undefined);
});

test('#convertRoman should convert string to utf-8 symbols', t => {
	Object.keys(cod).forEach(num => {
		t.is(arabicRoman.convertRoman(num), charToCode(num));
	});
});

test('#convertRoman should convert symbols not in dict', t => {
	t.is(arabicRoman.convertRoman('XV'), charToCode('X', 'V'));
	t.is(arabicRoman.convertRoman('XXX'), charToCode('X', 'X', 'X'));
	t.is(arabicRoman.convertRoman('CXXVIII'), charToCode('C', 'X', 'X', 'VIII'));
});

test('#convertRoman should support lowercase input', t => {
	t.is(arabicRoman.convertRoman('vii'), charToCode('VII'));
});


test/index-test.js
(function () {

    'use strict';

    mocha.setup('bdd');
    var assert = chai.assert;

    describe('ArabicRoman', function () {
        /**
         */
        var module = function () {
            return window.ArabicRoman;
        };

        var cod = {
            I: '\u2160',
            II: '\u2161',
            III: '\u2162',
            IV: '\u2163',
            V: '\u2164',
            VI: '\u2165',
            VII: '\u2166',
            VIII: '\u2167',
            IX: '\u2168',
            X: '\u2169',
            XI: '\u216A',
            XII: '\u216B',
            L: '\u216C',
            C: '\u216D',
            D: '\u216E',
            M: '\u216F',
        };
        /**
         * @return {String}
         */
        var charToCode = function () {
            var res = '';
            var args = arguments;
            var i;
            for (i = 0; i < args.length; i++) {
                res += cod[args[i]];
            }
            return res;
        };

        describe('constructor', function () {
            it('should initialize', function () {
                var m = module();
                assert.isDefined(m);
            });
        });

        describe('#toRoman', function () {
            it('should not convert number < 1', function () {
                var m = module();
                assert.isUndefined(m.toRoman(0));
                assert.isUndefined(m.toRoman(-1));
            });

            it('should not convert nubmer > 4000', function () {
                var m = module();
                assert.isDefined(m.toRoman(4000));
                assert.isUndefined(m.toRoman(4001));
            });

            it('should parse string attributes', function () {
                var m = module();
                assert.equal(m.toRoman('1'), '\u2160');
                assert.isUndefined(m.toRoman(NaN));
                assert.isUndefined(m.toRoman('a1'));
            });

            it('should convert arabic to roman numerals (from 1 to 10)', function () {
                var m = module();
                var codes = [
                    '',
                    charToCode('I'),
                    charToCode('II'),
                    charToCode('III'),
                    charToCode('IV'),
                    charToCode('V'),
                    charToCode('VI'),
                    charToCode('VII'),
                    charToCode('VIII'),
                    charToCode('IX'),
                    charToCode('X'),
                ];
                for (var i = 1; i < codes.length - 1; i++) {
                    assert.equal(m.toRoman(i), codes[i]);
                }
            });

            it('should convert big numbers', function () {
                var m = module();
                assert.equal(m.toRoman(11), charToCode('XI'));
                assert.equal(m.toRoman(12), charToCode('XII'));
                assert.equal(m.toRoman(50), charToCode('L'));
                assert.equal(m.toRoman(100), charToCode('C'));
                assert.equal(m.toRoman(500), charToCode('D'));
                assert.equal(m.toRoman(1000), charToCode('M'));
            });

            it('should convert complex numbers', function () {
                var m = module();
                assert.equal(m.toRoman(13), charToCode('X', 'III'));
                assert.equal(m.toRoman(14), charToCode('X', 'IV'));
                assert.equal(m.toRoman(15), charToCode('X', 'V'));
                assert.equal(m.toRoman(19), charToCode('X', 'IX'));
                assert.equal(m.toRoman(20), charToCode('X', 'X'));
                assert.equal(m.toRoman(39), charToCode('X', 'X', 'X', 'IX'));
                assert.equal(m.toRoman(4000), charToCode('M', 'M', 'M', 'M'));
                assert.equal(
                    m.toRoman(3888),
                    charToCode('M', 'M', 'M', 'D', 'C', 'C', 'C', 'L', 'X', 'X', 'X', 'VIII')
                );
                assert.equal(m.toRoman(3999), charToCode('M', 'M', 'M', 'C', 'M', 'X', 'C', 'IX'));
            });

            it('should convert hard numbers', function () {
                var m = module();
                assert.equal(m.toRoman(4), charToCode('IV'));
                assert.equal(m.toRoman(9), charToCode('IX'));
                assert.equal(m.toRoman(40), charToCode('X', 'L'));
                assert.equal(m.toRoman(90), charToCode('X', 'C'));
                assert.equal(m.toRoman(400), charToCode('C', 'D'));
                assert.equal(m.toRoman(900), charToCode('C', 'M'));
            });
        });
    });

    if (window.mochaPhantomJS) {
        mochaPhantomJS.run();
    } else {
        mocha.run();
    }

})();


index.js
/*!
 * arabic-roman-convert.js
 * Copyright(c) 2016 Vladimir Rodkin <mail@vovanr.com>
 * MIT Licensed
 */

(function (root, factory) {
	/* eslint-disable */
	if (typeof define === 'function' && define.amd) {
		define([], function () {
			return (root.ArabicRoman = factory());
		});
	} else if (typeof module === 'object' && module.exports) {
		module.exports = factory();
	} else {
		root.ArabicRoman = factory();
	}
	/* eslint-enable */
})(this, function () {
	'use strict';

	var arabic = [
		1000,
		900, // hard number
		500,
		400, // hard number
		100,
		90, // hard number
		50,
		40, // hard number
		12,
		11
	];
	var stringRoman = [
		'M',
		'#',
		'D',
		'#',
		'C',
		'#',
		'L',
		'#',
		'XII',
		'XI',
		'X',
		'IX',
		'VIII',
		'VII',
		'VI',
		'V',
		'IV',
		'III',
		'II',
		'I'
	];
	var roman = [
		'\u216F',
		'\u216D\u216F', // hard number
		'\u216E',
		'\u216D\u216E', // hard number
		'\u216D',
		'\u2169\u216D', // hard number
		'\u216C',
		'\u2169\u216C', // hard number
		'\u216B',
		'\u216A'
	];

	// from 10 to 1
	var i = 11;
	while (--i) {
		arabic.push(i);
		roman.push(String.fromCharCode(8543 + i));
	}

	var length = roman.length;

	/**
	 * @param {String} source
	 * @return {String}
	 */
	function convertRoman(source) {
		var result = '';
		var i;
		var current;
		var re;
		for (i = 0; i < length; i += 1) {
			current = stringRoman[i];
			re = new RegExp('^' + current, 'i');
			if (re.test(source)) {
				result = roman[i];
				result += convertRoman(source.substr(current.length, source.length));
				break;
			}
		}

		return result;
	}

	return {
		/**
		 * @param {Number|String} source
		 * @return {String}
		 */
		toRoman: function (source) {
			var result;

			if (
				(typeof source === 'string' && parseInt(source, 10) > 0) ||
				typeof source === 'number'
			) {
				result = this.arabicToRoman(source);
			} else {
				result = this.convertRoman(source);
			}

			return result;
		},

		/**
		 * Convert arabic nubmer to roman UTF-8 numerals
		 *
		 * @param {Number|String} number
		 * @return {String}
		 */
		arabicToRoman: function (number) {
			if (typeof number === 'string') {
				number = parseInt(number, 10);
			}

			if (typeof number !== 'number' || isNaN(number)) {
				return;
			}

			if (number <= 0 || number > 4000) {
				return;
			}

			var result = '';
			var i;
			var currentArabic;
			for (i = 0; i < length; i += 1) {
				currentArabic = arabic[i];
				while (number >= currentArabic && number > 0) {
					if (
						(currentArabic === 12 && number > 12) ||
						(currentArabic === 11 && number > 11)
					) {
						break;
					}
					result += roman[i];
					number -= currentArabic;
				}
			}

			return result;
		},

		/**
		 * Convert simple roman string to roman UTF-8 numerals
		 *
		 * @param {String} source
		 * @return {String}
		 */
		convertRoman: function (source) {
			if (typeof source !== 'string') {
				return;
			}

			if (!source.match(/^[ivxlcdm]+$/i)) {
				return;
			}

			return convertRoman(source);
		}
	};
});
